layers/igmp.go

```diff
@@ -44,26 +227,42 @@ func igmpTimeDecode(t uint8) time.Duration {
 	return time.Millisecond * 100 * time.Duration((mant|0x10)<<(exp+3))
 }
 
-// DecodeFromBytes decodes the given bytes into this layer.
-func (i *IGMP) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+// LayerType returns LayerTypeIGMP for the V1,2,3 message protocol formats.
+func (i *IGMP) LayerType() gopacket.LayerType       { return LayerTypeIGMP }
+func (i *IGMPv1Orv2) LayerType() gopacket.LayerType { return LayerTypeIGMP }
+
+func (i *IGMPv1Orv2) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
 	i.Type = IGMPType(data[0])
 	i.MaxResponseTime = igmpTimeDecode(data[1])
 	i.Checksum = binary.BigEndian.Uint16(data[2:4])
 	i.GroupAddress = net.IP(data[4:8])
-	if i.Type == 0x11 && len(data) > 8 {
-		i.SupressRouterProcessing = data[8]&0x8 != 0
-		i.RobustnessValue = data[8] & 0x7
-		i.IntervalTime = igmpTimeDecode(data[9])
-		numSources := int(binary.BigEndian.Uint16(data[10:12]))
-		for j := 0; j < numSources; j++ {
-			i.SourceAddresses = append(i.SourceAddresses, net.IP(data[12+j*4:16+j*4]))
-		}
-	} else {
-		i.SupressRouterProcessing = false
-		i.RobustnessValue = 0
-		i.IntervalTime = 0
-		i.SourceAddresses = nil
+
+	return nil
+}
+
+func (i *IGMPv1Orv2) NextLayerType() gopacket.LayerType {
+	return gopacket.LayerTypeZero
+}
+
+func (i *IGMPv1Orv2) CanDecode() gopacket.LayerClass {
+	return LayerTypeIGMP
+}
+
+// DecodeFromBytes decodes the given bytes into this layer.
+func (i *IGMP) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+
+	// common IGMP header values between versions 1..3 of IGMP specification..
+	i.Type = IGMPType(data[0])
+
+	switch i.Type {
+
+	case IGMPMembershipQuery:
+		i.decodeIGMPv3MembershipQuery(data)
+
+	case IGMPMembershipReportV3:
+		i.decodeIGMPv3MembershipReport(data)
 	}
```

Comment:
default to returning an error ("unsupported IGMP type", or something like that).